### PR TITLE
NODE-1860: Block Reward and Community Driven Monetary Policy

### PR DESCRIPTION
**Pull request by Karasiq**

### CODE REVIEW

File: node/src/main/scala/com/wavesplatform/database/Keys.scala [hunk 1]
```diff
@@ -166,4 +166,6 @@ object Keys {
   val InvokeScriptResultPrefix: Short = 56
   def invokeScriptResult(height: Int, txNum: TxNum): Key[InvokeScriptResult] =
     Key("invoke-script-result", hNum(InvokeScriptResultPrefix, height, txNum), InvokeScriptResult.fromBytes, InvokeScriptResult.toBytes)
+
+  val blockReward: Key[Option[Long]] = Key.opt("block-reward",  Array[Byte](0, 57), Longs.fromByteArray, Longs.toByteArray)
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/database/Keys.scala]: Please add this key to `Explorer`'s list of known keys

File: node/src/main/scala/com/wavesplatform/mining/Miner.scala [hunk 2]
```diff
@@ -124,10 +124,14 @@ class MinerImpl(allChannels: ChannelGroup,
       .leftMap(_.toString)
   }
 
+  private def getVersion(height: Int): Byte =
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/mining/Miner.scala]:
```
height <= blockVersion3AfterHeight: V2
!isFeatureActivated(BlockReward)  : V3
isFeatureActivated(BlockReward)   : V4
```

File: node/src/main/scala/com/wavesplatform/mining/microblocks/MicroBlockMinerImpl.scala [hunk 3]
```diff
@@ -128,13 +128,14 @@ class MicroBlockMinerImpl(debugState: Ref[Task, MinerDebugInfo.State],
       for {
         signedBlock <- Block
           .buildAndSign(
-            version = 3,
+            version = 4,
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/mining/microblocks/MicroBlockMinerImpl.scala]: Block version 4 should only be used after block reward feature is activated!

File: node/src/main/scala/com/wavesplatform/settings/FeaturesSettings.scala [hunk 4]
```diff
@@ -1,3 +1,22 @@
 package com.wavesplatform.settings
 
-case class FeaturesSettings(autoShutdownOnUnsupportedFeature: Boolean, supported: List[Short])
+import com.typesafe.config.Config
+import net.ceedubs.ficus.Ficus._
+import net.ceedubs.ficus.readers.ValueReader
+
+case class FeaturesSettings(autoShutdownOnUnsupportedFeature: Boolean, supported: List[Short], reward: Byte)
+
+object FeaturesSettings {
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/settings/FeaturesSettings.scala]: Is this really needed?

File: node/src/main/scala/com/wavesplatform/state/diffs/BlockDiffer.scala [hunk 5]
```diff
@@ -89,35 +90,47 @@ object BlockDiffer extends ScorexLogging {
...
-                                                          prevBlockTimestamp: Option[Long],
-                                                          prevBlockFeeDistr: Option[Portfolio],
-                                                          currentBlockFeeDistr: Option[Portfolio],
-                                                          txs: Seq[Transaction],
-                                                          verify: Boolean): TracedResult[ValidationError, Result[Constraint]] = {
+  private[this] def createDiff[Constraint <: MiningConstraint](blockchain: Blockchain,
+                                                               initConstraint: Constraint,
+                                                               prevBlockTimestamp: Option[Long],
+                                                               prevBlockFeeDistr: Option[Portfolio],
+                                                               currentBlockFeeDistr: Option[Portfolio],
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/state/diffs/BlockDiffer.scala]: `currentBlockFeeDistr` should already include block reward

File: node/src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 6]
```diff
@@ -425,6 +427,44 @@ class LevelDBWriter(private[database] val writableDB: DB,
...
+        import settings.functionalitySettings.blockRewardSettings._
+
+        val mayBeReward     = rw.get(Keys.blockReward(height))
+        val mayBeTimeToVote = height - activatedAt
+
+        mayBeReward match {
+          case Some(reward) if mayBeTimeToVote > 0 && mayBeTimeToVote % rewardPeriod == 0 =>
+            val votes       = blockRewardVotes(height - 1)
+            val currentVote = block.rewardVote
+            Some(((votes :+ currentVote).filter(_ >= 0) , reward))
```
Comment (chepiov) [inline node/src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: Filtering to settings.minReward?

File: node/src/main/scala/com/wavesplatform/state/diffs/BlockDiffer.scala [hunk 7]
```diff
@@ -37,29 +37,38 @@ object BlockDiffer extends ScorexLogging {
 
     // height switch is next after activation
     val ngHeight          = blockchain.featureActivationHeight(BlockchainFeatures.NG.id).getOrElse(Int.MaxValue)
+    val rewardHeight      = blockchain.featureActivationHeight(BlockchainFeatures.BlockReward.id).getOrElse(Int.MaxValue)
     val sponsorshipHeight = Sponsorship.sponsoredFeesSwitchHeight(blockchain)
 
-    lazy val prevBlockFeeDistr: Option[Portfolio] =
+    val minerRewardDistr: Portfolio =
+      if (stateHeight < rewardHeight)
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/state/diffs/BlockDiffer.scala]:
```suggestion
      if (blockchain.isFeatureActivated(BlockchainFeatures.BlockReward))
```

File: node/src/main/scala/com/wavesplatform/features/FeatureProvider.scala [hunk 8]
```diff
@@ -17,6 +18,11 @@ object FeatureProvider {
       else if (provider.approvedFeatures.get(feature).exists(_ <= height)) BlockchainFeatureStatus.Approved
       else BlockchainFeatureStatus.Undefined
 
+    def currentBlockVersion: Byte =
+      if (isFeatureActivated(BlockchainFeatures.BlockReward)) RewardBlockVersion
+      else if (isFeatureActivated(BlockchainFeatures.NG)) NgBlockVersion
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/features/FeatureProvider.scala]: NG block is activated at specific height, before all other features.

File: node/src/test/scala/com/wavesplatform/history/package.scala [hunk 9]
```diff
@@ -19,7 +19,8 @@ package object history {
   val DefaultBlockchainSettings = BlockchainSettings(
     addressSchemeCharacter = 'N',
     functionalitySettings = TestFunctionalitySettings.Enabled,
-    genesisSettings = null
+    genesisSettings = null,
+    rewardsSettings = null
```
Comment (phearnot) [inline node/src/test/scala/com/wavesplatform/history/package.scala]: Why not some "default" value?

File: node/src/test/scala/com/wavesplatform/mining/MiningWithRewardSuite.scala [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+import com.wavesplatform.account.{Address, KeyPair}
+import com.wavesplatform.block.Block
+import com.wavesplatform.common.state.ByteStr
+import com.wavesplatform.common.utils._
+import com.wavesplatform.consensus.PoSSelector
+import com.wavesplatform.database.{Keys, LevelDBWriter}
+import com.wavesplatform.db.DBCacheSettings
+import com.wavesplatform.features.BlockchainFeatures
+import com.wavesplatform.lagonaki.mocks.TestBlock
+import com.wavesplatform.settings.{
```
Comment (phearnot) [inline node/src/test/scala/com/wavesplatform/mining/MiningWithRewardSuite.scala]:
```suggestion
import com.wavesplatform.settings._
```

File: node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala [hunk 11]
```diff
@@ -124,6 +124,45 @@ class BlockchainUpdaterImpl(private val blockchain: LevelDBWriter,
...
+          case Some(reward) if mayBeTimeToVote > 0 && mayBeTimeToVote % settings.term == 0 =>
+            Some((votes , reward))
+          case None if mayBeTimeToVote == 0 =>
+            Some((Seq(), settings.initial))
+          case _ => None
+        }
+      }
+      .flatMap {
+        case (votes, currentReward) =>
+          val lt = votes.count(_ < currentReward).toDouble / settings.votingInterval
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala]: Please use `Int`s/`Long`s here. No need to convert to `Double`s.  Vote count should just be greater than or equal to `settings.votingInterval / 2 + 1`.

File: node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala [hunk 12]
```diff
@@ -124,6 +124,45 @@ class BlockchainUpdaterImpl(private val blockchain: LevelDBWriter,
     }
   }
 
+  private def rewardForBlock(block: Block): Option[Long] = {
```
Comment (phearnot) [inline node/src/main/scala/com/wavesplatform/state/BlockchainUpdaterImpl.scala]: Shouldn't this just return a `Long`?

**Pull request merged**